pallets/author-height-filter/src/lib.rs

```diff
@@ -0,0 +1,172 @@
+//! disfavor authors who are authoring a disproportionate amount of the time in an attempt to
+//! "even the playing field".
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::pallet;
+
+pub use pallet::*;
+
+#[pallet]
+pub mod pallet {
+
+	use log::debug;
+	use frame_support::pallet_prelude::*;
+	use frame_support::traits::Randomness;
+	use sp_std::vec::Vec;
+	use frame_system::pallet_prelude::*;
+	use sp_core::H256;
+	use sp_runtime::Percent;
+
+	/// The Author Filter pallet
+	#[pallet::pallet]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	/// Configuration trait of this pallet.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type
+		type Event: From<Event> + IsType<<Self as frame_system::Config>::Event>;
+		/// Deterministic on-chain pseudo-randomness used to do the filtering
+		type RandomnessSource: Randomness<H256, Self::BlockNumber>;
+		/// A source for the complete set of potential authors.
+		/// The starting point of the filtering.
+		type PotentialAuthors: Get<Vec<Self::AccountId>>;
+	}
+
+	// This code will be called by the author-inherent pallet to check whether the reported author
+	// of this block is eligible at this height. We calculate that result on demand and do not
+	// record it instorage (although we do emit a debugging event for now).
+	impl<T: Config> pallet_author_inherent::CanAuthor<T::AccountId> for Pallet<T> {
+		fn can_author(account: &T::AccountId) -> bool {
+			// Use 0 as a bogus relay block.
+			Self::can_author_helper(account, 0)
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Helper method to calculate eligible authors
+		/// This function still takes _relay_height as a parameter even though it won't be used
+		/// This is ground work to help me understand what needs abstracted out into a trait.
+		pub fn can_author_helper(account: &T::AccountId, _relay_height: u32) -> bool {
+			let mut active: Vec<T::AccountId> = T::PotentialAuthors::get();
+
+			let num_eligible = EligibleRatio::<T>::get().mul_ceil(active.len());
+			let mut eligible = Vec::with_capacity(num_eligible);
+
+			for i in 0..num_eligible {
+				// A context identifier for grabbing the randomness. Consists of two parts
+				// - The constant string *b"filter" - to identify this pallet
+				// - The index `i` when we're selecting the ith eligible author
+				// This has the weakness that the authors are based only on para-block height.
+				// For a more robust solution that adds entropy from the parachain inherent, see
+				// ../author-slot-filter
+				let subject: [u8; 7] = [b'f', b'i', b'l', b't', b'e', b'r', i as u8];
+				let (randomness, _) = T::RandomnessSource::random(&subject);
+
+				// TODO why could we not use the same method as in moonbeam?
+				// Why does to_low_u64_be not exist on H256?
+				// https://docs.rs/primitive-types/0.9.0/primitive_types/struct.H256.html#method.to_low_u64_be
+				// Cast to u32 first so we get consistent results on 32- and 64-bit platforms.
```

Comment:
what 64-bit platforms?

Reply:
At first we were just casting directly to `usize` which is 32-bit in wasm which most authoring nodes were using. The problem was full nodes that ran with native execution couldn't import the blocks because they were getting a different index.